File: Python/GraphTheory/LCA.py
class LCA:
    def __init__(self, g, root=0):
        self.root = root
        self.g = g
        self.depth = [-1 if i != self.root else 0 for i in range(len(self.g))]
        self.uku = (len(self.g) - 1).bit_length()
        self.par = [[-1] * len(self.g) for _ in range(self.uku)]
        que = [self.root]
        while que:
            u = que.pop()
            for i in self.g[u]:
                if self.depth[i] == -1:
                    self.depth[i] = self.depth[u] + 1
                    self.par[0][i] = u
                    que += [i]

        for i in range(1, self.uku):
            for j in range(len(self.g)):
                if self.par[i - 1][j] != -1:
                    self.par[i][j] = self.par[i - 1][self.par[i - 1][j]]

    def query(self, u, v):
        if self.depth[v] < self.depth[u]:
            u, v = v, u
        for i in range(self.uku):
            if self.depth[v] - self.depth[u] >> i & 1:
                v = self.par[i][v]
        if u == v:
            return u
        for i in range(self.uku - 1, -1, -1):
            uu, vv = self.par[i][u], self.par[i][v]
            if uu != vv:
                u, v = uu, vv
        return self.par[0][u]

'''verify: https://onlinejudge.u-aizu.ac.jp/problems/GRL_5_C
n = int(input())
g = [[] for _ in range(n)]
for i in range(n):
    li = list(map(int, input().split()))
    for j in range(li[0]):
        g[i] += [li[j + 1]]

l = LCA(g)
q = int(input())
for _ in range(q):
    u, v = map(int, input().split())
    print(l.query(u, v))
'''

File: C++/GraphTheory/LCA.cpp
struct LCA {
	int root;
	const int n, dep;
	vector<vector<int>> parent;
	vector<int> depth;

	LCA(const vector<vector<int>>& graph, int root, int n)
		: root(root), n(n), dep(log2(n) + 1), parent(dep, vector<int>(n)), depth(n) {
		dfs(graph, root, root, 1);
		for (int k = 0; k + 1 < dep; ++k) {
			parent.push_back(vector<int>(n, root));
			for (int v = n - 1; v >= 0; --v) {
				int par = parent[k][v];
				parent[k + 1][v] = parent[k][par];
			}
		}
	}

	void dfs(const vector<vector<int>>& graph, int from, int par, int d) {
		parent[0][from] = par;
		depth[from] = d;
		for (int to : graph[from]) {
			if (depth[to] > 0) continue;
			dfs(graph, to, from, d + 1);
		}
	}

	int query(int v1, int v2) {
		if (depth[v1] < depth[v2]) swap(v1, v2);
		for (int k = 0; k < dep; ++k) {
			if ((depth[v1] - depth[v2]) >> k & 1) {
				v1 = parent[k][v1];
			}
		}
		if (v1 == v2) { return v1; }
		for (int k = dep - 1; k >= 0; --k) {
			int p_v1 = parent[k][v1];
			int p_v2 = parent[k][v2];
			if (p_v1 == p_v2) continue;
			v1 = p_v1;
			v2 = p_v2;
		}
		assert(parent[0][v1] == parent[0][v2]);
		return parent[0][v1];
	}
};

/*verify: https://onlinejudge.u-aizu.ac.jp/problems/GRL_5_C
int main() {
	int n, q;
	cin >> n;
	vector<vector<int>>gr(1e5);
	for (int i = 0; i < n; ++i) {
		int k, c;
		cin >> k;
		for (int j = 0; j < k; ++j) {
			cin >> c;
			gr[i].push_back(c);
		}
	}
	cin >> q;
	LCA lca = LCA(gr, 0, n);
	for (int i = 0; i < q; ++i) {
		int u, v;
		cin >> u >> v;
		cout << lca.query(u, v) << endl;
	}
}
*/